docs/source/format/CDataInterface/PyCapsuleInterface.rst

```diff
@@ -0,0 +1,402 @@
+                A PyCapsule containing a C ArrowSchema representation of a requested
+                schema. Conversion to this schema is best-effort.
+
+            Returns
+            -------
+            Tuple[PyCapsule, PyCapsule]
+                A pair of PyCapsules containing a C ArrowSchema and ArrowArray,
+                respectively.
+            """
+            ...
+
+
+    class ArrowStreamExportable(Protocol):
+        def __arrow_c_stream__(self, requested_schema: object | None = None) -> object:
+            """
+            Get a PyCapsule containing a C ArrowArrayStream representation of the object.
+
+            The capsule will have a name of "arrow_array_stream".
+
+            If requested_schema is passed, the callee should attempt to provide the
+            data in the requested schema. However, this is best-effort, and the
+            callee may return a PyCapsule containing an ArrowArray with a different
+            schema. This parameter is useful for cases where the underlying data
+            could be represented in multiple ways, and the caller has a preference
+            for how it is represented. For example, some systems have a single
+            integer type, but Arrow has multiple integer types with different 
+            sizes and sign.
+
+            Parameters
+            ----------
+            requested_schema : PyCapsule | None
+                A PyCapsule containing a C ArrowSchema representation of a requested
+                schema. Conversion to this schema is best-effort.
+
+            Returns
+            -------
+            PyCapsule
+                A PyCapsule containing a C ArrowArrayStream representation of the
+                object.
+            """
+            ...
+
+Schema Requests
+---------------
+
+In some cases, there might be multiple possible Arrow representations of the
+same data. For example, a library might have a single integer type, but Arrow
+has multiple integer types with different sizes and sign. As another example,
+Arrow has several possible encodings for an array of strings: 32-bit offsets,
+64-bit offsets, string view, and dictionary-encoded. A sequence of strings could
+export to any one of these Arrow representations.
+
+In order to allow the caller to request a specific representation, the
+``__arrow_c_array__`` and ``__arrow_c_stream__`` methods may take an optional
+``requested_schema`` parameter. This parameter is a PyCapsule containing an
+``ArrowSchema``.
+
+The callee should attempt to provide the data in the requested schema. However,
+if the callee cannot provide the data in the requested schema, they may return
+with the same schema as if ``None`` were passed to ``requested_schema``.
+
+If the caller requests a schema that is not compatible with the data,
+say requesting a schema with a different number of fields, the callee should
+raise an exception. The requested schema mechanism is only meant to negotiate
```

Comment:
```suggestion
raise an exception, preferably ``TypeError`` or a subclass thereof.
The requested schema mechanism is only meant to negotiate
```

Reply:
"preferably" is of course not strong, so maybe fine to keep, but I am not sure we generally should expect a single class of error. There are multiple reasons why a certain cast is not possible, for example because the types are completely incompatible (TypeError), or because a specific value in the array cannot be cast (ValueError).

Reply:
Hmm, that's true. I withdraw my suggestion then.